src/main/java/com/algorithmia/algorithmHandler/AlgorithmHandler.java

```diff
@@ -0,0 +1,107 @@
+package com.algorithmia.algorithmHandler;
+
+import java.lang.reflect.Method;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+
+public class AlgorithmHandler<INPUT, OUTPUT, STATE> {
+
+    private ReflectionHelper.DebuggableBifunction<INPUT, STATE, OUTPUT> applyWState;
+    private ReflectionHelper.DebuggableFunction<INPUT, OUTPUT> apply;
+    private Supplier<STATE> loadFunc;
+
+    private Class algorithmClass;
+    private STATE state;
+
+
+    public AlgorithmHandler(Class algorithmClass, ReflectionHelper.DebuggableBifunction<INPUT, STATE, OUTPUT> applyWState, Supplier<STATE> loadFunc) {
+        this.applyWState = applyWState;
+        this.loadFunc = loadFunc;
+        this.algorithmClass = algorithmClass;
+    }
+
+    public AlgorithmHandler(Class algorithmClass, ReflectionHelper.DebuggableBifunction<INPUT, STATE, OUTPUT> applyWState) {
+        this.applyWState = applyWState;
+        this.algorithmClass = algorithmClass;
+    }
+
+    public AlgorithmHandler(Class algorithmClass, ReflectionHelper.DebuggableFunction<INPUT, OUTPUT> apply) {
+        this.apply = apply;
+        this.algorithmClass = algorithmClass;
+    }
+
+    private void Load() {
+        if (this.loadFunc != null) {
+            state = this.loadFunc.get();
+            System.out.println("PIPE_INIT_COMPLETE");
+            System.out.flush();
+        }
+    }
+
+    private void ExecuteWithoutState(RequestHandler<INPUT> in, ResponseHandler out, Function<INPUT, OUTPUT> func) {
+        INPUT req = in.GetNextRequest();
+        while (req != null) {
+            OUTPUT output = func.apply(req);
+            out.writeToPipe(output);
+            req = in.GetNextRequest();
+        }
+    }
+
+    private void ExecuteWithState(RequestHandler<INPUT> in, ResponseHandler out, BiFunction<INPUT, STATE, OUTPUT> func) {
```

Comment:
also the idea of passing in a state object feels a little unnecessary since the Algorithms have to be an object anyway, and  during the `load` function could instantiate whatever it needs instead of needing to get that object passed in every time anyway...

Reply:
The underlying Handlers have some components that are not threadsafe, so passing an object around I think is a good move. However I think they could easily be class variables instanced at construction instead of local variables.

Reply:
do you mean the actual usercode that would be getting executed? i thought they'd only be processing a single thing at a time anyway though, and also it wasn't clear to me whether the `executeWithState` is given a mutable or immutable reference to the state object and if it's mutable then you'd have concurrency problems there anyway... basically the user is creating a state object that their own code wants access to, so I don't really see why it's getting copied around instead of just letting the user instantiate whatever fields they need in their load method?

Reply:
Yeah they would be processing 1 at a time, sequentially - however, creating a new ResponseHandler and Requesthandler aren't free, and if something doesn't get released properly we may have multiple RW references to the same underlying file. I don't have any concrete examples of how that could happen, I'm spitballing a potential problem.

The current solution should pre-empt anything around this (check the latest commit for modifications).

Reply:
WRT state, mutable state is permitted in the contract however it's not going to be strictly enforced. The objective here is to provide immutable state to algorithms at runtime.